src/runtime/hexagon_host.cpp

```diff
@@ -0,0 +1,643 @@
+
+    #ifdef DEBUG_RUNTIME
+    uint64_t t_before = halide_current_time_ns(user_context);
+    #endif
+
+    debug(user_context) << "    ion_alloc len=" << (uint64_t)size << ", heap_id=" << heap_id << " -> ";
+    int ion_fd;
+    void *ion = ion_alloc(user_context, size, heap_id, &ion_fd);
+    debug(user_context) << "        " << ion << ", " << ion_fd << "\n";
+    if (!ion) {
+        error(user_context) << "ion_alloc failed\n";
+        return -1;
+    }
+
+    if (remote_register_buf) {
+        // If we have this symbol, it means that we can register the
+        // buffer to get zero copy behavior with the DSP.
+        debug(user_context) << "    remote_register_buf buf=" << ion << " size=" << (int)size << " fd=" << ion_fd << "\n";
+        remote_register_buf(ion, size, ion_fd);
+    }
+
+    int err = halide_hexagon_wrap_device_handle(user_context, buf, ion, size);
+    if (err != 0) {
+        ion_free(user_context, ion);
+        return err;
+    }
+
+    if (!buf->host) {
+        // If the host pointer has also not been allocated yet, set it to
+        // the ion buffer. This buffer will be zero copy.
+        buf->host = (uint8_t *)ion;
+        debug(user_context) << "    host <- " << buf->host << "\n";
+    }
+
+    #ifdef DEBUG_RUNTIME
+    uint64_t t_after = halide_current_time_ns(user_context);
+    debug(user_context) << "    Time: " << (t_after - t_before) / 1.0e6 << " ms\n";
+    #endif
+
+    return 0;
+}
+
+WEAK int halide_hexagon_device_free(void *user_context, buffer_t* buf) {
+    debug(user_context)
+        << "Ion: halide_hexagon_device_free (user_context: " << user_context
+        << ", buf: " << buf << ")\n";
+
+    #ifdef DEBUG_RUNTIME
+    uint64_t t_before = halide_current_time_ns(user_context);
+    #endif
+
+    void *ion = halide_hexagon_detach_device_handle(user_context, buf);
+    ion_free(user_context, ion);
+
+    if (remote_register_buf) {
+        // If we have this symbol, we registered the buffer, so now we
+        // unregister it (by setting fd = -1).
+        size_t size = buf_size(user_context, buf);
+        debug(user_context) << "    remote_register_buf buf=" << ion << " size=" << (int)size << " fd=-1\n";
+        remote_register_buf(ion, size, -1);
+    }
+
+    if (buf->host == ion) {
+        // If we also set the host pointer, reset it.
+        buf->host = NULL;
+        debug(user_context) << "    host <- 0x0\n";
+    }
+
+    #ifdef DEBUG_RUNTIME
+    uint64_t t_after = halide_current_time_ns(user_context);
+    debug(user_context) << "    Time: " << (t_after - t_before) / 1.0e6 << " ms\n";
+    #endif
+
+    return 0;
+}
+
+namespace {
+
+// Implement a device copy using memcpy.
+WEAK void device_memcpy(void *user_context, device_copy c) {
+    if (c.src == c.dst) {
+        // This is a zero copy buffer, copy is a no-op.
+        return;
+    }
+    // TODO: Is this 32-bit or 64-bit? Leaving signed for now
```

Comment:
The copies get simplified before they're used: https://github.com/halide/Halide/blob/master/src/runtime/cuda_opencl_shared.h#L96